### DISCUSSION TITLE
[BUG] RustDesk server listening on IPv6 sockets, and not listening on IPv4, on Debian with no IPv6 address

### DISCUSSION THREAD

**Discussion by bc-109** [rustdesk/rustdesk-server#535] [General]
Hi,
I made a fresh install of hbbs / hbbr on a Debian 12 VM. IPv6 is disabled. The VM only has one IPv4 address :

`ip a`
`1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
`
`2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:15:5d:00:02:b8 brd ff:ff:ff:ff:ff:ff
    inet 10.111.222.20/24 brd 10.111.222.255 scope global eth0
       valid_lft forever preferred_lft forever`

When running hbbs, it seems to start listening on IPv6 sockets (that's funny on a machine with no IPv6 address ), and it does not listen at all on IPv4 sockets :

`netstat -an | grep 2111`
`tcp6       0      0 :::21117                :::*                    LISTEN`     
`tcp6       0      0 :::21116                :::*                    LISTEN `    
`tcp6       0      0 :::21119                :::*                    LISTEN  `   
`tcp6       0      0 :::21118                :::*                    LISTEN  `   
`tcp6       0      0 :::21115                :::*                    LISTEN  `   
`udp6       0      0 :::21116                :::* `

Moreover, I didn't find any configuration file / parameter where I could specify on which IP address I would like it to bind.

As it does not listen at all on IPv4 sockets, of course, it does not accept connections via IPv4 !

Thank you in advance

**Comment by bc-109** [comment]
I tried to open a bug, but it has been closed with no explanation :
#7056 
Most server software allow to specify in configuration file how to bind it to a specific IP address. I found the command-line switch to change the listening port, but nothing to change the listening IP address.

I agree, this may not be a bug, this may be a "lack of feature" (c) :-)

For now, hbbs does not listen on my IPv4 address. So, it's pretty useless...

**Comment by bc-109** [comment]
Hi,

Bug #7056 is still closed, I have no idea why.

In the meantime, a friend of mine made a new install on different VM, on different location, with different OS (Ubuntu), and with default network configuration (both IPv6 and IPv6 addresses).

Same weird behavior :
- In UDP, the server listens in IPv4 and IPv6
- In TCP, it listens only in IPv6

As only IPv4 is available on the WAN link and on the gateway router, this leads to the following behavior :
- On LAN, all works perfectly
- On WAN, the client connects to the server (through UDPv4) but it's impossible to take control over another machine (as the server does not listen on TCPv4).

The faulty code seems to be here. I don't really know Rust language, but I would say if IPv6 socket succeeds, it does not ever try to open an IPv4 socket :
https://github.com/rustdesk/rustdesk/blob/1f557888f5a61991d86928a6ab814dd21ac82b56/libs/hbb_common/src/tcp.rs#L279C1-L286C10

As this behavior has been seen on two different locations, on two different VMs and OSes, on two different installations made by two different people, I can't imagine we're the only guys facing that problem...

Does that mean RustDesk is IPv6 only, and IPv4 / dual stack is not supported at all ?

KInd regards

**Comment by monat78** [comment]
Same thing, and there's no way to tell it to listen on ipv4, it only listens on ipv6, use debian 12 distro.

**Comment by monat78** [comment]
version: '3'

networks:
  rustdesk-net:
    external: false

services:
  hbbs:
    container_name: hbbs
    ports:
      - 21115:21115
      - 21116:21116
      - 21116:21116/udp
      - 21118:21118
    image: rustdesk/rustdesk-server:latest
    command: hbbs -r sub.domain.com:21117
    volumes:
      - ./data:/root
    networks:
      - rustdesk-net
    depends_on:
      - hbbr
    restart: unless-stopped

  hbbr:
    container_name: hbbr
    ports:
      - 21117:21117
      - 21119:21119
    image: rustdesk/rustdesk-server:latest
    command: hbbr
    volumes:
      - ./data:/root
    networks:
      - rustdesk-net
    restart: unless-stopped
~

**Comment by bc-109** [comment]
Hi @monat78. 

Please, could you explain the previous post with some words ? This post seems to show some data about hbbs / hbbr running in containers, and it does not show any distinction between IPv4 and IPv6. My bug report was about running hbbs / hbbr on bare Linux machines (Debian and Ubuntu), without any container. What I said is that when hbbs binds to IPv6, it does not ever try to bind to IPv4. And, a sub-problem is that it manages to bind on IPv6 even on machines where there's no IPv6 ! 

This is not a correct behavior. "Dual stack" good practices, where both IPv4 and IPv6 are active, is to bind on both protocols for incoming connections. For outgoing connections, good practices are to try IPv6 first, and if it fails, try IPv4.

I browsed the source, and I pointed out a piece of code where I think the problem resides. Anyway, I'm not a Rust programmer, and I can't solve the problem by mysef.

Does your answer mean RustDesk does not support IPv4 ? Or that RustDesk is supported only inside containers ? 

Thank you in advance. Kind regards

**Comment by rogerjames99** [comment]
I have encountered exactly the same problem on ubuntu 22.04. So I forked a copy of the rustdesk-server repository on github and did some tinkering in the code. I added some diagnostics to the new_socket functions in tcp.rs and udp.rs. I also have zero knowledge of rust! However I managed to cobble together something that built.

This showed that only ipv6 addresses were being asked for! This means that the problem must be in the server default config.

The config mechanism seems to be pretty much undocumented. The only thing I could find was some environment variables in the the top level REAMDE.md.

My next step is to set RUST_LOG=trace and see what I can find.

**Comment by rogerjames99** [comment]
Here are the mods.
[diffs.txt](https://github.com/rustdesk/rustdesk/files/14814986/diffs.txt)

Here is the trace log.
[rustlog.txt](https://github.com/rustdesk/rustdesk/files/14814993/rustlog.txt)

I still have not got as clue as to why no ipv4 addresses are being requested. I think the answer my lie somewhere in socket_client.rs.

Rust expertise urgently required!

**Reply by rogerjames99** [reply]
I now tracked the problem to these two files.

```diff --git a/src/rendezvous_server.rs b/src/rendezvous_server.rs
index 0c0d107..af3eab3 100644
--- a/src/rendezvous_server.rs                                    +++ b/src/rendezvous_server.rs
@@ -1296,7 +1296,7 @@ async fn create_udp_listener(port: i32, rmem: usize) -> ResultType<FramedSocket>
     let addr = SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), port as _);
     if let Ok(s) = FramedSocket::new_reuse(&addr, false, rmem).await {
         log::debug!("listen on udp {:?}", s.local_addr());
-        return Ok(s);
+    //    return Ok(s);                                               }
     let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), port as _);
     let s = FramedSocket::new_reuse(&addr, false, rmem).await?;
```
  This patch fixes the udp ports and seems to work.

```diff --git a/libs/hbb_common/src/tcp.rs b/libs/hbb_common/src/tcp.rs
index f574e83..0c6be48 100644
--- a/libs/hbb_common/src/tcp.rs
+++ b/libs/hbb_common/src/tcp.rs
@@ -282,8 +282,10 @@ pub async fn listen_any(port: u16) -> ResultType<TcpListener> {
             .bind(SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), port))
             .is_ok()
         {
-            if let Ok(l) = socket.listen(DEFAULT_BACKLOG) {
-                return Ok(l);
+            if let Ok(_l) = socket.listen(DEFAULT_BACKLOG) {
+            /*
+               return Ok(l); 
+            */
             }
```   
This patch compiles but results in ipv6 being disabled. My non-existent knowledge of rust means I haven't a clue on how to fix it.

The only way forward I can see is to raise this a pull request and ask for a code review.

**Comment by xlionjuan** [comment]
You'd better try deploy though docker

And the reason why your issues got closed is because you should post your issues at discussions, issues only store organized issues about devs.

And you said:

> Bug https://github.com/rustdesk/rustdesk/issues/7056 is still closed, I have no idea why.

I hope you look carefully.

https://github.com/rustdesk/rustdesk/issues/7056#issuecomment-1928765056

**Comment by rogerjames99** [comment]
I have created pull request [#395](https://github.com/rustdesk/rustdesk-server/pull/395)

**Comment by rogerjames99** [comment]
I also tried the work around of using the latest docker images. These exhibit exactly the same problem as the bare metal version. IPv6 ports only!

**Comment by alatark** [comment]
I get the same erroneous behaviour on a newly-created VM (the current standard Ubuntu 22.04 image used by Windows 11 Hyper-V Quick Create) and just following the instructions at https://rustdesk.com/docs/en/self-host/rustdesk-server-oss/install/

Clearly it's not working as intended...

**Reply by xlionjuan** [reply]
Please build server on Linux, and docker is more recommend and more easy

**Reply by alatark** [reply]
> Please build server on Linux, and docker is more recommend and more easy

Both the OP and myself **are** on Linux (Debian and Ubuntu respectively)!

In my case, it is admittedly the version of Ubuntu 22.04 LTS that's supplied thru Windows Hyper-V, but it's still standard 22.04 LTS. I wonder if it doesn't have something to do with them trying to push IPv6 by deprecating some old IPv4 defaults. I might try on 20.04 LTS instead.

I'll try the docker version as well...

**Reply by xlionjuan** [reply]
I have no problem on both IPv6/IPv4 on both Ubuntu 22.04/24.04(testing) and Arch

And all of them is using docker s6 image

**Reply by alatark** [reply]
Thanks, that's helpful. If my theory (see below) is correct, then if you run

`netstat -an | grep 2111 ` 

you will only see TCP6 listeners (as per the original post).

My theory is that it is listening on IPv4 too, even though no IPv4 listeners are listed by netstat.

In other words, there may never have been a problem, only a quirk in the netstat utility that makes it look like there is.

**Reply by alatark** [reply]
Update: my system is working properly -- and thus clearly listening -- on IPv4 despite netstat showing that it is only listening on IPv6.

**Comment by alatark** [comment]
By my count that makes at least five of us with the same issue (bc109, their friend, monat76, rogerjames109, and myself).

There's clearly some systematic issue.

**Comment by alatark** [comment]
Perhaps the issue is not that IPv4 isn't enabled, just that netstat doesn't report them as listening. I note that a number of people have issues with confusion about nginx (appearing in netstat to be) not listening on IPv4, even when it actually is.

Using ss (rather than the deprecated netstat) also shows the ports as listening with -lp and with -lp6, but not with -lp4:
```
root@rustdesk-test:~# ss -lp | grep "hbb"
udp   UNCONN 0      0                                                       *:53814                           *:*    users:(("hbbs",pid=534,fd=16))                                                             
udp   UNCONN 0      0                                                       *:21116                           *:*    users:(("hbbs",pid=534,fd=12))                                                             
tcp   LISTEN 0      128                                                     *:21115                           *:*    users:(("hbbs",pid=534,fd=14))                                                             
tcp   LISTEN 0      128                                                     *:21118                           *:*    users:(("hbbs",pid=534,fd=15))                                                             
tcp   LISTEN 0      128                                                     *:21119                           *:*    users:(("hbbr",pid=533,fd=10))                                                             
tcp   LISTEN 0      128                                                     *:21116                           *:*    users:(("hbbs",pid=534,fd=13))                                                             
tcp   LISTEN 0      128                                                     *:21117                           *:*    users:(("hbbr",pid=533,fd=9))      
```                                                        
```
root@rustdesk-test:~# ss -lp6 | grep "hbb"
udp   UNCONN 0      0                  *:53814                *:*    users:(("hbbs",pid=534,fd=16))          
udp   UNCONN 0      0                  *:21116                *:*    users:(("hbbs",pid=534,fd=12))          
tcp   LISTEN 0      128                *:21115                *:*    users:(("hbbs",pid=534,fd=14))          
tcp   LISTEN 0      128                *:21118                *:*    users:(("hbbs",pid=534,fd=15))          
tcp   LISTEN 0      128                *:21119                *:*    users:(("hbbr",pid=533,fd=10))          
tcp   LISTEN 0      128                *:21116                *:*    users:(("hbbs",pid=534,fd=13))          
tcp   LISTEN 0      128                *:21117                *:*    users:(("hbbr",pid=533,fd=9))   
```        
```
root@rustdesk-test:~# ss -lp4 | grep "hbb"
root@rustdesk-test:~# 
```

However, an external nmap scan of the box shows the ports responding to IPv4 ack scans:

```
Host is up, received arp-response (0.0010s latency).
Scanned at 2024-04-07 15:08:41 Eastern Daylight Time for 1s

PORT      STATE      SERVICE REASON
21110/tcp filtered   unknown no-response
21111/tcp filtered   unknown no-response
21112/tcp filtered   unknown no-response
21113/tcp filtered   unknown no-response
21114/tcp filtered   unknown no-response
21115/tcp unfiltered unknown reset ttl 64
21116/tcp unfiltered unknown reset ttl 64
21117/tcp unfiltered unknown reset ttl 64
21118/tcp unfiltered unknown reset ttl 64
21119/tcp unfiltered unknown reset ttl 64
21120/tcp filtered   unknown no-response

```

**Comment by alatark** [comment]
Update: my system is working properly -- and thus clearly listening -- on IPv4 despite netstat showing that it is only listening on IPv6. Thus there never was an issue, only the appearance of one.

I'm not entirely clear on why this is, perhaps IPv6 automatically deals with the IPv4 address space using the ::ffff:0:0/96 special address range that maps IPv4 addresses.

In any case, I found several other software projects (e.g. nginx) which note a similar behaviour.

**Comment by rogerjames99** [comment]
I have withdrawn my pull request. It was nonsensical! My only excuse is that I have been recovering from a bout of pneumonia and have lost the plot somewhat. The code is working as expected. When you create an ipv6socket, set the ipv6_only flag to false, and bind to the the ipv6 "any" address "[::]:*". The socket will accept ipv4 connections, but it will map the incoming ipv4 address into an ipv6 address using the standard rules. Here is an example.

```
roger@dragon:~$ sudo netstat -lp|grep debug
tcp6       0      0 [::]:21115              [::]:*                  LISTEN      16258/target/debug/ 
tcp6       0      0 [::]:21116              [::]:*                  LISTEN      16258/target/debug/ 
tcp6       0      0 [::]:21118              [::]:*                  LISTEN      16258/target/debug/ 
udp6       0      0 [::]:21116              [::]:*                              16258/target/debug/ 
udp6       0      0 [::]:43860              [::]:*                              16258/target/debug/ 
roger@dragon:~$ telnet 10.151.0.2 21115
Trying 10.151.0.2...
Connected to 10.151.0.2.
Escape character is '^]'.
^C^[^]
telnet> q
Connection closed.
roger@dragon:~$
```

**Comment by aa51513** [comment]
Unfortunately, I also encountered the same problem. 

I don’t want to hear too much theoretical knowledge. I just want to say that I can access through `[::1]:port`, but cannot access through `127.0.0.1:port`.

**Reply by noname1001** [reply]
You can force ip4 by adapting your `docker-compose.yml`

for example

```yml
services:

  hbbs:
    container_name: rustdesk-hbbs
    image: rustdesk/rustdesk-server:latest
    command: hbbs
    volumes:
      - ./data:/root
    ports:
      # 21114 TCP web console, only available in Pro version
      # 21115 TCP for NAT type test
      # 21116 UDP/TCP TCP Hole Punching / Heartbeat / ID-Server
      # 21118 TCP Relay
      - "192.168.0.200:21114-21116:21114-21116/tcp"
      - "192.168.0.200:21118:21118/tcp"
      - "192.168.0.200:21116:21116/udp"
    depends_on:
      - hbbr
    restart: always

  hbbr:
    container_name: rustdesk-hbbr
    image: rustdesk/rustdesk-server:latest
    command: hbbr
    volumes:
      - ./data:/root
    ports:
      # 21117 TCP Relay
      # 21119 TCP Relay
      - "192.168.0.200:21117:21117/tcp"
      - "192.168.0.200:21119:21119/tcp"
    restart: always
```

For me it works.

In case you want to know more here is the [official documentation](https://rustdesk.com/docs/en/self-host/rustdesk-server-oss/synology/dsm-7/) which helped me.

**Reply by guitarbug** [reply]
> You can force ip4 by adapting your `docker-compose.yml`
> 
> for example
> 
> ```yaml
> services:
> 
>   hbbs:
>     container_name: rustdesk-hbbs
>     image: rustdesk/rustdesk-server:latest
>     command: hbbs
>     volumes:
>       - ./data:/root
>     ports:
>       # 21114 TCP web console, only available in Pro version
>       # 21115 TCP for NAT type test
>       # 21116 UDP/TCP TCP Hole Punching / Heartbeat / ID-Server
>       # 21118 TCP Relay
>       - "192.168.0.200:21114-21116:21114-21116/tcp"
>       - "192.168.0.200:21118:21118/tcp"
>       - "192.168.0.200:21116:21116/udp"
>     depends_on:
>       - hbbr
>     restart: always
> 
>   hbbr:
>     container_name: rustdesk-hbbr
>     image: rustdesk/rustdesk-server:latest
>     command: hbbr
>     volumes:
>       - ./data:/root
>     ports:
>       # 21117 TCP Relay
>       # 21119 TCP Relay
>       - "192.168.0.200:21117:21117/tcp"
>       - "192.168.0.200:21119:21119/tcp"
>     restart: always
> ```
> 
> For me it works.
> 
> In case you want to know more here is the [official documentation](https://rustdesk.com/docs/en/self-host/rustdesk-server-oss/synology/dsm-7/) which helped me.

it's not working for my rustdesk on ubuntun...

**Reply by xlionjuan** [reply]
Check natstat


If you still don't know why, reinstall OS

**Comment by jamesyc** [comment]
I had the same issue. 

Easily fixed it by removing `network_mode: "host"` from the docker compose config, and setting the ports manually. 

```services:
  hbbs:
    container_name: rustdesk-hbbs
    image: rustdesk/rustdesk-server:latest
    command: hbbs
    volumes:
      - ./data:/root
    ports:
      # 21114 TCP web console, only available in Pro version
      # 21115 TCP for NAT type test
      # 21116 UDP/TCP TCP Hole Punching / Heartbeat / ID-Server
      # 21118 TCP Relay
      - "0.0.0.0:21114-21116:21114-21116/tcp"
      - "0.0.0.0:21118:21118/tcp"
      - "0.0.0.0:21116:21116/udp"
    depends_on:
      - hbbr
    restart: unless-stopped

  hbbr:
    container_name: rustdesk-hbbr
    image: rustdesk/rustdesk-server:latest
    command: hbbr
    volumes:
      - ./data:/root
    ports:
      # 21117 TCP Relay
      # 21119 TCP Relay
      - "0.0.0.0:21117:21117/tcp"
      - "0.0.0.0:21119:21119/tcp"
    restart: unless-stopped```

**Reply by daveymg-nz** [reply]
Same issue still occurring on self hosted docker install, this port config resolved it for me, thanks.
